tutorial/lesson_21_auto_scheduler_generate.cpp

```diff
@@ -0,0 +1,192 @@
+// Halide tutorial lesson 21: Auto-Scheduler
+
+// This lesson demonstrates how to use the auto-scheduler to generate a
+// copy-pastable CPU schedule that can be subsequently improved upon.
+
+// On linux or os x, you can compile and run it like so:
+
+// g++ lesson_21_auto_scheduler_generate.cpp ../tools/GenGen.cpp -g -std=c++11 -fno-rtti -I ../include -L ../bin -lHalide -lpthread -ldl -o lesson_21_generate
+// export LD_LIBRARY_PATH=../bin   # For linux
+// export DYLD_LIBRARY_PATH=../bin # For OS X
+// ./lesson_21_generate -o . -f conv_layer target=host
+// g++ lesson_21_auto_scheduler_run.cpp brighten_*.o -ldl -lpthread -o lesson_21_run
+// ./lesson_21_run
+
+// If you have the entire Halide source tree, you can also build it by
+// running:
+//    make tutorial_lesson_21_auto_scheduler_run
+// in a shell with the current directory at the top of the halide
+// source tree.
+
+#include "Halide.h"
+#include <stdio.h>
+
+using namespace Halide;
+
+// We will define a generator to auto-schedule.
+class AutoScheduled : public Halide::Generator<AutoScheduled> {
+public:
+    GeneratorParam<bool>  auto_schedule{"auto_schedule", false};
+
+    Input<Buffer<float>>  input{"input", 4};
+    Input<Buffer<float>>  filter{"filter", 4};
+    Input<Buffer<float>>  bias{"bias", 1};
+    Input<float>          min_value{"min_value"};
+
+    Output<Buffer<float>> output1{"output1", 4};
+    Output<Buffer<float>> output2{"output2", 4};
+
+    void generate() {
+        RDom r(filter.dim(0).min(), filter.dim(0).extent(),
+               filter.dim(1).min(), filter.dim(1).extent(),
+               filter.dim(2).min(), filter.dim(2).extent());
+
+        f(x, y, z, n) = bias(z);
+        f(x, y, z, n) += filter(r.x, r.y, r.z, z) * input(x + r.x, y + r.y, r.z, n);
+        output1(x, y, z, n) = max(0, f(x, y, z, n));
+        output2(x, y, z, n) = min(min_value, f(x, y, z, n));
+    }
+
+    void schedule() {
+        if (auto_schedule) {
+            // To use the auto-scheduler, we need to provide estimates on all
```

Comment:
Replace this para with: 

The auto-scheduler requires estimates on all the input/output sizes and parameter values in order to compare different alternatives and decide on a good schedule.